control-plane/csi-driver/controller/src/pvwatcher.rs

```diff
@@ -0,0 +1,145 @@
+        watcher(pv_handle, ListParams::default())
+            .touched_objects()
+            .try_for_each(|pvol| async {
+                process_object(pvol).await;
+                Ok(())
+            })
+            .await
+            .unwrap();
+        Ok(())
+    }
+}
+
+///Accepts PV object received from event stream and applies Garbage collector logic.
+async fn process_object(pv: PersistentVolume) {
+    let client = Client::try_default().await.unwrap();
+    let pv_handle = Api::<PersistentVolume>::all(client.clone());
+    let mut deletion_candidate = false;
+    let mut our_pv = false;
+    match pv.metadata.clone().deletion_timestamp {
+        Some(_) => deletion_candidate = true,
+        None => {}
+    }
+
+    if let Some(provisioner) = &pv.spec.as_ref().unwrap().csi {
+        if provisioner.driver == MAYASTOR_PLUGIN {
+            our_pv = true;
+        }
+    }
+
+    if deletion_candidate && our_pv {
+        if let Some(reclaim_policy) = &pv.spec.as_ref().unwrap().persistent_volume_reclaim_policy {
+            if let Some(phase) = &pv.status.as_ref().unwrap().phase {
+                if reclaim_policy == "Retain" && phase == "Released" {
+                    debug!(
+                        "PV :{} is a deletion candidate",
+                        &pv.metadata.clone().name.unwrap()
+                    );
+                    let vol_handle: &String = &pv
+                        .spec
+                        .as_ref()
+                        .unwrap()
+                        .csi
+                        .as_ref()
+                        .unwrap()
+                        .volume_handle;
+                    delete_volume(vol_handle).await;
+                }
+                if phase == "Bound" {
+                    if let Ok(pvol) = pv_handle.get_opt(&pv.metadata.clone().name.unwrap()).await {
+                        match pvol {
+                            Some(_) => debug!(
+                                "PV {} present on API server",
+                                &pv.metadata.clone().name.unwrap()
+                            ),
+                            None => {
+                                debug!(
+                                    "PV :{} is a deletion candidate",
+                                    &pv.metadata.clone().name.unwrap()
+                                );
+                                let vol_handle: &String = &pv
+                                    .spec
+                                    .as_ref()
+                                    .unwrap()
+                                    .csi
+                                    .as_ref()
+                                    .unwrap()
+                                    .volume_handle;
+                                delete_volume(vol_handle).await;
+                            }
+                        }
+                    } else {
+                        debug!("Unknown error");
```

Comment:
Can we print the actual error rather than the generic statement? You can do a match on line no 76 and you will get the error object in the error arm.

Reply:
Will do

Reply:
Incorporated